sel/src/Sel/SecretKey/EncryptedStream.hs

```diff
@@ -0,0 +1,650 @@
+            (Foreign.castPtr additionalDataPtr)
+            (fromIntegral additionalDataLength)
+            tag
+
+-- This functions is meant to be called either from 'pushToStream'.
+doPushToStream
+  :: forall (s :: Type)
+   . Multipart s
+  -- ^ the cryptographic state
+  -> StrictByteString
+  -- ^ Message
+  -> Ptr CUChar
+  -- ^ Additional data pointer
+  -> CULLong
+  -- ^ Additional data length
+  -> StreamTag
+  -- ^ Stream tag
+  -> IO (Either EncryptedStreamError CipherText)
+doPushToStream (Multipart statePtr) message additionalDataPointer additionalDataLength tag =
+  BS.unsafeUseAsCStringLen message $ \(cString, cStringLen) -> do
+    let cipherTextLength = fromIntegral cryptoSecretStreamXChaCha20Poly1305ABytes + cStringLen
+    cipherTextFPtr <- Foreign.mallocForeignPtrBytes cipherTextLength
+    Foreign.withForeignPtr cipherTextFPtr $ \cipherTextBuffer -> do
+      result <-
+        cryptoSecretStreamXChaCha20Poly1305Push
+          statePtr
+          cipherTextBuffer
+          Foreign.nullPtr
+          (Foreign.castPtr @CChar @CUChar cString)
+          (fromIntegral @Int @CULLong cStringLen)
+          additionalDataPointer
+          additionalDataLength
+          (streamTagToConstant tag)
+      case result of
+        0 -> do
+          pure $ Right $ CipherText cipherTextFPtr (fromIntegral cipherTextLength)
+        _ -> pure $ Left EncryptionStreamPushError
+
+-- | Provide a cryptographic context 'Multipart' in a continuation to encrypt a stream.
+--
+-- @since 0.0.1.0
+encryptStream
+  :: forall (m :: Type -> Type)
+   . MonadIO m
+  => (forall (s :: Type). Multipart s -> IO [CipherText])
+  -- ^ Continuation in which the stream gets encrypted
+  -> m (Header, SecretKey, [CipherText])
+encryptStream action = do
+  (SecretKey secretKeyFPtr) <- liftIO newSecretKey
+  headerFPtr <- liftIO $ Foreign.mallocForeignPtrBytes (fromIntegral cryptoSecretStreamXChaCha20Poly1305HeaderBytes)
+  liftIO $ Foreign.withForeignPtr headerFPtr $ \headerPtr ->
+    liftIO $ Foreign.withForeignPtr secretKeyFPtr $ \keyPtr -> do
+      Foreign.allocaBytes (fromIntegral cryptoSecretStreamXChaCha20Poly1305StateBytes) $ \statePtr -> do
+        cryptoSecretStreamXChaCha20Poly1305InitPush statePtr headerPtr keyPtr
+        result <- action $ Multipart statePtr
+        pure (Header headerFPtr, SecretKey secretKeyFPtr, result)
+
+-- == Decryption ==
+
+-- | Initialise a stream from which you will retrieve encrypted messages.
+--
+-- You need the 'Header' and 'SecretKey' used by your peer to intialise the stream in order to
+-- sucecessfully decrypt the incoming messages.
+--
+-- @since 0.0.1.0
+initPullStream
+  :: Multipart s
+  -- ^ Cryptographic state.
+  -> Header
+  -- ^ Header received from the peer, used to ensure the ciphertexts will be valid.
+  -> SecretKey
+  -- ^ Secret key from the peer, used to decrypt the ciphertexts.
+  -> IO (Either EncryptedStreamError ())
+  -- ^ Returns 'True' on success, 'False' if the header is invalid
```

Comment:
Was this meant to return an `IO (Either EncryptedStreamError Bool)`?

Reply:
Actually the doc is wrong, I'm experimenting with more expressive error types.